iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift

```diff
@@ -0,0 +1,185 @@
+    private var insertedIndexPaths: Set<IndexPath> = []
+
+    override func prepare(forCollectionViewUpdates updateItems: [UICollectionViewUpdateItem]) {
+        super.prepare(forCollectionViewUpdates: updateItems)
+
+        guard let collectionView = self.collectionView,
+        let mostRecentOffset = self.getMostRecentItemContentOffset() else { return }
+
+        for update in updateItems {
+            switch update.updateAction {
+            case .insert:
+                guard let indexPath = update.indexPathAfterUpdate else { break }
+                self.insertedIndexPaths.insert(indexPath)
+
+                let isScrolledToMostRecent = (mostRecentOffset.y - collectionView.contentOffset.y) <= self.itemHeight
+                // Always scroll to the end for new user messages, or if we're currently scrolled to the
+                // most recent message.
+                if indexPath.section == 1 || isScrolledToMostRecent {
+                    self.shouldScrollToEnd = true
+                }
+            case .delete:
+                guard let indexPath = update.indexPathBeforeUpdate else { break }
+                self.deletedIndexPaths.insert(indexPath)
+            case .reload, .move, .none:
+                break
+            @unknown default:
+                break
+            }
+        }
+    }
+
+    override func finalizeCollectionViewUpdates() {
+        super.finalizeCollectionViewUpdates()
+
+        self.shouldScrollToEnd = false
+        self.deletedIndexPaths.removeAll()
+        self.insertedIndexPaths.removeAll()
+    }
+
+    override func targetContentOffset(forProposedContentOffset proposedContentOffset: CGPoint) -> CGPoint {
+        guard self.shouldScrollToEnd, let offset = self.getMostRecentItemContentOffset() else {
+            return super.targetContentOffset(forProposedContentOffset: proposedContentOffset)
+        }
+
+        return offset
+    }
+
+    override func finalLayoutAttributesForDisappearingItem(at itemIndexPath: IndexPath)
+    -> UICollectionViewLayoutAttributes? {
+        guard deletedIndexPaths.contains(itemIndexPath) else { return nil }
+
+        return super.finalLayoutAttributesForDisappearingItem(at: itemIndexPath)
+    }
+
+    override func initialLayoutAttributesForAppearingItem(at itemIndexPath: IndexPath)
+    -> UICollectionViewLayoutAttributes? {
+
+        let attributes = super.initialLayoutAttributesForAppearingItem(at: itemIndexPath)
+
+        // A new message dropped in the user's message stack should appear immediately.
+        guard itemIndexPath.section == 1 else {
+            return attributes
+        }
+
+        // Ensure this is actually a new message and not just a placeholder message.
+        if self.insertedIndexPaths.contains(itemIndexPath)
+            && self.dataSource?.getTimeMachineItem(forItemAt: itemIndexPath) != nil {
+            attributes?.alpha = 1
+        }
+
+        return attributes
+    }
+}
```

Comment:
The TimeMachineCollectionViewLayout diff might not show because it's too large. TLDR is that it's pretty much the same as before but less of it.